bin/propolis-server/src/lib/server.rs

```diff
@@ -332,6 +337,79 @@ async fn register_oximeter(
     Ok(registry)
 }
 
+/// Wrapper around a [`NexusClient`] object, which allows deferring
+/// the DNS lookup until accessed.
+///
+/// Without the assistance of OS-level DNS lookups, the [`NexusClient`]
+/// interface requires knowledge of the target service IP address.
+/// For some services, like Nexus, this can be painful, as the IP address
+/// may not have even been allocated when the Sled Agent starts.
+///
+/// This structure allows clients to access the client on-demand, performing
+/// the DNS lookup only once it is actually needed.
+struct Inner {
+    log: Logger,
+    resolver: Resolver,
+}
+#[derive(Clone)]
+pub struct LazyNexusClient {
+    inner: Arc<Inner>,
+}
+
+impl LazyNexusClient {
+    pub fn new(log: Logger, addr: Ipv6Addr) -> Result<Self, ResolveError> {
+        Ok(Self {
+            inner: Arc::new(Inner {
+                log: log.clone(),
+                resolver: Resolver::new_from_ip(log, addr)?,
+            }),
+        })
+    }
+
+    pub async fn get_ip(&self) -> Result<SocketAddrV6, ResolveError> {
+        self.inner.resolver.lookup_socket_v6(ServiceName::Nexus).await
+    }
+
+    pub async fn get(&self) -> Result<NexusClient, ResolveError> {
+        let address = self.get_ip().await?;
+
+        Ok(NexusClient::new(
+            &format!("http://{}", address),
+            self.inner.log.clone(),
+        ))
+    }
+}
+
+// Use our local address as basis for calculating a Nexus endpoint,
+// Return that endpoint if successful.
+async fn find_local_nexus_client(
+    local_addr: SocketAddr,
+    log: Logger,
+) -> Option<NexusClient> {
+    // At the moment, we only support converting an IPv6 address into a
+    // Nexus endpoint.
+    let address = match local_addr {
+        SocketAddr::V6(my_address) => *my_address.ip(),
+        SocketAddr::V4(_) => return None,
```

Comment:
Perhaps worthwhile emitting a `warn` message here as well, to match the other cases where we won't be emitting a usable `NexusClient`?

Reply:
Done